numpyro/contrib/module.py

```diff
@@ -429,3 +431,381 @@ def random_haiku_module(
+            call_kwargs["rngs"] = rngs
+
+        result = model(*call_args, **call_kwargs)
+
+        # Update mutable state if needed
+        if mutable and module_state:
+            for state_type in module_state:
+                for path in list(module_state[state_type].keys()):
+                    if "." in path:
+                        parent, child = path.split(".", 1)
+                        if hasattr(model, parent):
+                            parent_obj = getattr(model, parent)
+                            if hasattr(parent_obj, child):
+                                if hasattr(getattr(parent_obj, child), "value"):
+                                    module_state[state_type][path] = getattr(
+                                        parent_obj, child
+                                    ).value
+                                else:
+                                    module_state[state_type][path] = getattr(
+                                        parent_obj, child
+                                    )
+
+        return result
+
+    return apply_fn
+
+
+def batchnorm_in_module(module_cls):
+    """Helper function to check if a module contains BatchNorm layers."""
+    import inspect
+
+    source = inspect.getsource(module_cls)
+    return "BatchNorm" in source
+
+
+def random_nnx_module(
+    name,
+    nn_module,
+    prior,
+    *args,
+    input_shape=None,
+    apply_rng=None,
+    mutable=None,
+    **kwargs,
+):
+    """
+    A primitive to create a random :mod:`~flax.nnx` style neural network
+    which can be used in MCMC samplers. The parameters of the neural network
+    will be sampled from ``prior``.
+
+    :param str name: name of the module to be registered.
+    :param flax.nnx.Module nn_module: a `flax nnx` Module which follows the NNX API
+    :param prior: a distribution or a dict of distributions or a callable.
+        If it is a distribution, all parameters will be sampled from the same
+        distribution. If it is a dict, it maps parameter names to distributions.
+        If it is a callable, it takes parameter name and parameter shape as
+        inputs and returns a distribution.
+    :param args: optional arguments to initialize NNX neural network
+        as an alternative to `input_shape`
+    :param tuple input_shape: shape of the input taken by the
+        neural network.
+    :param list apply_rng: A list to indicate which extra rng _kinds_ are needed for
+        ``nn_module``. Defaults to None, which means no extra rng key is needed.
+    :param list mutable: A list to indicate mutable states of ``nn_module``. For example,
+        if your module has BatchNorm layer, we will need to define ``mutable=["batch_stats"]``.
+    :param kwargs: optional keyword arguments to initialize NNX neural network
+        as an alternative to `input_shape`
+    :return: a callable that takes an array as an input and returns
+        the neural network transformed output array.
+    """
```

Comment:
I think you can reuse `_update_params` together with https://flax.readthedocs.io/en/latest/guides/surgery.html as in `random_haiku_module`.